catalyst/data/sampler_inbatch.py

```diff
@@ -0,0 +1,199 @@
+            labels: labels of the samples in the batch
+
+        Returns: indeces of the batch samples to forming triplets.
+        """
+        raise NotImplementedError
+
+    def sample(self, features: Tensor, labels: List[int]) -> TTriplets:
+        """
+        Args:
+            features: has the shape of [batch_size, feature_size]
+            labels: labels of the samples in the batch
+
+        Returns:
+            the batch of the triplets in the order below:
+            (anchor, positive, negative)
+        """
+        self._check_input_labels(labels=labels)
+
+        ids_anchor, ids_pos, ids_neg = self._sample(
+            features=features, labels=labels
+        )
+
+        return features[ids_anchor], features[ids_pos], features[ids_neg]
+
+
+class AllTripletsSampler(InBatchTripletsSampler):
+    """
+    This sampler selects all the possible triplets for the given labels
+    """
+
+    def __init__(self, max_output_triplets: int = maxsize):
+        """
+        Args:
+            max_output_triplets: with the strategy of choosing all
+                the triplets, their number in the batch can be very large,
+                because of it we can sample only random part of them,
+                determined by this parameter.
+        """
+        self._max_out_triplets = max_output_triplets
+
+    def _sample(self, *_: Tensor, labels: List[int]) -> TTripletsIds:
+        """
+        Args:
+            labels: labels of the samples in the batch
+            *_: note, that we ignore features argument
+
+        Returns: indeces of triplets
+        """
+        num_labels = len(labels)
+
+        triplets = []
+        for label in set(labels):
+            ids_pos_cur = set(find_value_ids(labels, label))
+            ids_neg_cur = set(range(num_labels)) - ids_pos_cur
+
+            pos_pairs = list(combinations(ids_pos_cur, r=2))
+
+            tri = [(a, p, n) for (a, p), n in product(pos_pairs, ids_neg_cur)]
+            triplets.extend(tri)
+
+        triplets = sample(triplets, min(len(triplets), self._max_out_triplets))
+        ids_anchor, ids_pos, ids_neg = zip(*triplets)
+
+        return list(ids_anchor), list(ids_pos), list(ids_neg)
+
+
+class HardTrripletsSampler(InBatchTripletsSampler):
+    """
+    This sampler selects hardest triplets based on distances between features:
+    the hardest positive sample has the maximal distance to the anchor sample,
+    the hardest negative sample has the minimal distance to the anchor sample.
+
+    """
+
+    def __init__(self, need_norm: bool = False):
+        """
+        Args:
+            need_norm: set True if features normalisation is needed
+        """
+        self._need_norm = need_norm
+
+    def _sample(self, features: Tensor, labels: List[int]) -> TTripletsIds:
+        """
+        This method samples the hardest triplets inside the batch.
+
+        Args:
+            features: has the shape of [batch_size, feature_size]
```

Comment:
could we add annotations to the docs?

Reply:
I'm sorry, but I do not agree here. 
Because docs becomes outdated very fast and it is very hard to catch it.
At the same time we have typings which can be checked via mypy.